### PR TITLE
#5172 - Formatter actions

### LINKED ISSUE #5172: [APP] Formatter app and actions
**Name of app / service**
Formatter. The app that contain utilities for transforming date time, number and text

**Is lack of support preventing you from building workflows, or do you have a workaround?**
Using Node.js/Python code. First reported by a customer here: https://pipedream-users.slack.com/archives/CPTJYRY5A/p1673518421024549

**Actions specification**
## [Date / Time] Format

Format a date string to another date string

### Props

- **Input**:
    - **Description**: Date string. Pipedream will try to interpret this input, if the interpreted input is incorrect, please set your format in the `From Format` prop.
    - **Type:** string
    - **Required:** true
- **To Format:**
    - **Description:** Provide the format that the date is converted to
    - **Type:** Options
        - ddd MMM DD HH:mm:ss Z YYYY (Sun Jan 22 23:04:05 -0000 2006)
        - MMMM DD YYYY HH:mm:ss (January 22 2006 23:04:05)
        - MMMM DD YYYY (January 22 2006)
        - MMM DD YYYY (Jan 22 2006)
        - YYYY-MM-DDTHH:mm:ssZ (2006-01-22T23:04:05-0000)
        - YYYY-MM-DD HH:mm:ss Z (2006-01-22 23:04:05 -0000)
        - YYYY-MM-DD (2006-01-22)
        - MM-DD-YYYY (01-22-2006)
        - MM/DD/YYYY (01/22/2006)
        - MM/DD/YY (01/22/06)
        - DD-MM-YYYY (22-01-2006)
        - DD/MM/YYYY (22/01/2006)
        - DD/MM/YY (22/01/06)
        - X (1137971045)
    - **Required:** true
- **From Format:**
    - **Description:** Used when Pipedream incorrectly interpret the incoming (input) date, set this to explicitly tell Pipedream the format
    - **Type:** Options
        - ddd MMM DD HH:mm:ss Z YYYY (Sun Jan 22 23:04:05 -0000 2006)
        - MMMM DD YYYY HH:mm:ss (January 22 2006 23:04:05)
        - MMMM DD YYYY (January 22 2006)
        - MMM DD YYYY (Jan 22 2006)
        - YYYY-MM-DDTHH:mm:ssZ (2006-01-22T23:04:05-0000)
        - YYYY-MM-DD HH:mm:ss Z (2006-01-22 23:04:05 -0000)
        - YYYY-MM-DD (2006-01-22)
        - MM-DD-YYYY (01-22-2006)
        - MM/DD/YYYY (01/22/2006)
        - MM/DD/YY (01/22/06)
        - DD-MM-YYYY (22-01-2006)
        - DD/MM/YYYY (22/01/2006)
        - DD/MM/YY (22/01/06)
        - X (1137971045)
    - **Required:** false

### Output:

- The transformed date follow the `To Format` prop

## [Date / Time] Add / Subtract Time

Add or Subtract time

### Props:

- **Input**:
    - **Description**: Date you would like to manipulate in [[ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601)](https://en.wikipedia.org/wiki/ISO_8601), for example: `2023-01-13T02:58:12+00:00`
    - **Type:** string
    - **Required:** true
- **Operation**:
    - **Description:** The operation on the input
    - **Type:** Options with the following values:
        - Add
        - Subtract
- **Duration:**
    - **Description**: The duration for the operation. You can use the shorthand duration, for example: `1s` , `1m`, `1h`, `1d`, `1w`, `1y` equal one second, minute, hour, day, week, and year respectively
    - **Type**: string

### Output:

- The new date that produced by applying the operation (add/subtract) the input to the duration. The output should be in ISO 8601 format

## [Date / Time] Compare Dates

Get the duration between two dates in days, hours, minutes, and seconds along with checking if they are the same.

### Props:

- **Start Date Value:**
    - **Description:** Enter start date. If the start date is after the end date, these dates will be swapped and in the output "datesSwapped" will be set to True.
    - **Required:** true
- **Start Date Format:**
    - **Description:** Used when Pipedream incorrectly interpret the incoming (input) date, set this to explicitly tell Pipedream the format.
    - **Type:** Options
        - ddd MMM DD HH:mm:ss Z YYYY (Sun Jan 22 23:04:05 -0000 2006)
        - MMMM DD YYYY HH:mm:ss (January 22 2006 23:04:05)
        - MMMM DD YYYY (January 22 2006)
        - MMM DD YYYY (Jan 22 2006)
        - YYYY-MM-DDTHH:mm:ssZ (2006-01-22T23:04:05-0000)
        - YYYY-MM-DD HH:mm:ss Z (2006-01-22 23:04:05 -0000)
        - YYYY-MM-DD (2006-01-22)
        - MM-DD-YYYY (01-22-2006)
        - MM/DD/YYYY (01/22/2006)
        - MM/DD/YY (01/22/06)
        - DD-MM-YYYY (22-01-2006)
        - DD/MM/YYYY (22/01/2006)
        - DD/MM/YY (22/01/06)
        - X (1137971045)
    - **Required:** false
- **End Date:**
    - **Description:** Enter end date. Timezone is assumed the same for both dates if not explicitly set.
    - **Required:** true

### Output:

- An object with `result` field for the duration, and the `datesSwapped` field to tell if we have swapped the start date and end date. If the start date equals to the end date, the `result` should be `equal`
    - For example: `{ result: "1 day", datesSwapped: true}`
    - For example: `{ result: "equal", datesSwapped: false}`

## [Numbers] Format Currency

Format a number as a currency.

### Props:

- **Input:**
    - **Description:** Number you would like to format as a currency.
    - **Type:** number
    - **Required:** true
- **Currency:**
    - **Description: Specify the currency to be used for formatting**
    - **Type:** Options. Follow the [[available currency symbols here](https://en.wikipedia.org/wiki/Currency_symbol)](https://en.wikipedia.org/wiki/Currency_symbol)
    - **Required:** true
- **Currency Format:**
    - **Description:** Specify the format to be used for the currency formatting. Use the unicode currency symbol (¤) for special formatting options. Formatting rules can be found here: [http://www.unicode.org/reports/tr35/tr35-numbers.html#Number_Format_Patterns](http://www.unicode.org/reports/tr35/tr35-numbers.html#Number_Format_Patterns)
    - **Type:** Options.
        - ¤#,##0.00
        - ¤#,##0.00 ¤¤
        - ¤#,##0.00 ¤¤¤
        - ¤###0.00
        - #,##0.00
        - ###0.00
    - **Required:** true

### Output:

- The formatted string follow the Currency and Currency Format

## [Numbers] Format Number

Format a number to a new style. Does not perform any rounding or padding of the number.

### Props:

- **Input:**
    - **Description:** Number string you would like to format.
    - **Type:** string
    - **Required:** true
- **Input Decimal Mark:**
    - **Description:** The character the input uses to denote the decimal/fractional portion of the number.
    - **Type:** Options
        - Comma
        - Period
    - **Required:** false
- **To Format:**
    - **Description: The format the number will be converted to.**
    - **Type:** Options
        - Comma for grouping & period for decimal
        - Period for grouping & comma for decimal
        - Space for grouping & period for decimal
        - Space for grouping & comma for decimal
    - **Required:** true

### Output:

- The formatted string follow the `To Format` prop

### Props:

- **Input:**
    - **Description:** the text you would like to transform
    - **Type:** string
    - **Required:** true
- **Operation:**
    - **Description:** The case operation
    - **Type:** Options
        - **Capitalize**: Capitalize the first character of every word.
        - **Titlecase**: Convert all characters in a string to titlecase
        - **Pluralize**: Pluralize any English word (frog turns into frogs; child turns into children)
        - **Uppercase**: Capitalize every character in the text
        - **Lowercase:** Convert all characters in a string to lowercase
    - **Required:** true

### Output:

- The transformed text based on the `Operation`

## [Text] Convert HTML to Markdown

Convert valid HTML to Markdown text

### Props

- **Input:**
    - **Description:** HTML you would like to convert to Markdown.
    - **Type:** string
    - **Required:** true

### Output:

- The matching Markdown

## [Text] Convert Markdown to HTML

Convert Markdown text into valid HTML

### Props

- **Input:**
    - **Description:** Markdown text you would like to convert to HTML.
    - **Type:** string
    - **Required:** true

### Output:

- The matching HTML

## [Text] Default Value

Return a default value if the text is empty

### Props

- **Input:**
    - **Description:** Text you would like to check if empty.
    - **Type:** string
    - **Required:** true
- **Default Value:**
    - **Description:** Value to return if the text is empty.
    - **Type:** string
    - **Required:** true
    
    ### Output:
    
    - The default value if the text is empty, the text if the text is not empty
    

## [Text] Extract Email Address

Find an email address out of a text field. Finds the first email address only.

### Props

- **Input:**
    - **Description:** Text you would like to find an email address from.
    - **Type:** string
    - **Required:** true

### Output:

- The first email extracted from the text field if found. `null` value if not found

## [Text] Extract Number

Find a number in text. Finds the first number only.

### Props

- **Input:**
    - **Description:** Text you would like to find a number from.
    - **Type:** string
    - **Required:** true

### Output:

- The first number from the text field if found. `null` value if not found

## [Text] Extract by Regular Expression

Find a match for a regular expression pattern. Returns all matched groups with start and end position. Default is first match and no flags.

### Props

- **Input:**
    - **Description:** Text you would like to find a pattern from.
    - **Type:** string
    - **Required:** true
- **Regular Expression:**
    - **Description:** Enter a [[Javascript Regular Expression](https://www.w3schools.com/jsref/jsref_obj_regexp.asp)](https://www.w3schools.com/jsref/jsref_obj_regexp.asp) to find a match for, e.g. `/ab+c/i`.
    - **Type:** string
    - **Required:** true

### Output:

- A match for a regular expression pattern
- Returns all matched groups with start and end position
- Default is first match and no flags

## [Text] Extract Phone Number

Find and copy a complete phone number out of a text field. Finds the first phone number only.

### Props

- **Input:**
    - **Description:** Text you would like to find a phone number from.
    - **Type:** string
    - **Required:** true
- **Phone Number Format:**
    - **Description:** The phone number format to extract. If there's no format fit your needs, please [submit a ticket for Pipedeam](https://github.com/PipedreamHQ/pipedream/issues/new/choose), or use Extract by Regular Expression instead
    - **Type:** Options
        - North American Number Plan (NANP) e.g. (123) 456-7890
        - International e.g. (12) 34-56-78-90

### Output:

- The phone number matching the format

## [Text] Extract URL

Find and copy a web URL out of a text field. Finds the first URL only.

### Props

- **Input:**
    - **Description:** Text you would like to find a web URL from.
    - **Type:** string
    - **Required:** true

### Output:

- The found url

## [Text] Replace

Replace all instances of any character, word or phrase in the text with another character, word or phrase.

### Props

- **Input:**
    - **Description:** Text you would like to find and replace values within.
    - **Type:** string
    - **Required:** true
- **Find Text:**
    - **Description:** The text you would like to search for and replace, exact match.
    - **Type:** string
    - **Required:** false
- **Find Regular Expression:**
    - **Description:** Enter a [[Javascript Regular Expression](https://www.w3schools.com/jsref/jsref_obj_regexp.asp)](https://www.w3schools.com/jsref/jsref_obj_regexp.asp) to search for and replace, e.g. `/ab+c/i`.
    - **Type:** string
    - **Required:** false
- **Replace Text:**
    - **Description:** Leave blank to delete the found text
    - **Type:** string
    - **Required:** true

### Output:

- The input with the text replaced. `Find Text` or `Find Regular Expression` must be present

## [Text] Split Text

Split the text on a character or word and return a segment

### Props

- **Input:**
    - **Description:** String you would like to split.
    - **Type:** string
    - **Required:** true
- **Separator Text:**
    - **Description:** Character or word separator to split the text on
    - **Type:** string
    - **Required:** false
- **Separator Regular Expression:**
    - **Description:** Enter a [[Javascript Regular Expression](https://www.w3schools.com/jsref/jsref_obj_regexp.asp)](https://www.w3schools.com/jsref/jsref_obj_regexp.asp) to split, e.g. `/ab+c/i`.
    - **Type:** string
    - **Required:** false
- **Segment Index:**
    - **Description:** Segment of text to return after splitting
    - **Type:** Options
        - First
        - Second
        - Last
        - Second to Last
        - All
    - **Required:** true

### Output:

- The splitted segments. `Separator Text` or `Separator Regular Expression` must be present

## [Text] Trim Whitespace

Removes leading and trailing whitespace.

### Props

- **Input:**
    - **Description:** Text you would like to remove leading and trailing whitespace from.
    - **Type:** string
    - **Required:** true

### Output:

- The text with leading and trailing whitespace removed

## [Text] URL Decode

Decode the encoded URL, for example `https%3A%2F%2Fwww.google.com%2Fsearch%3Fq%3Dgeeks%2Bfor%2Bgeeks` becomes `https://www.google.com/search?q=geeks+for+geeks`

### Props

- **Input:**
    - **Description:** Text format you would like to decode.
    - **Type:** string
    - **Required:** true

### Output:

- The decoded url

## [Text] URL Encode

Encode the URL, for example `https://www.google.com/search?q=geeks+for+geeks` becomes `https%3A%2F%2Fwww.google.com%2Fsearch%3Fq%3Dgeeks%2Bfor%2Bgeeks` 

### Props

- **Input:**
    - **Description:** Text format you would like to decode.
    - **Type:** string
    - **Required:** true

### Output:

- The decoded url

### PR DESCRIPTION
**Pull request by GTFalcao**
19 new actions for the new `Formatting` utilities app.

Closes #5172

### CODE REVIEW

File: components/formatting/actions/format-number/format-number.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      label: "Input",
+      description: "Number string you would like to format.",
+      type: "string",
+    },
+    inputDecimalMark: {
+      label: "Input Decimal Mark",
+      description:
+        "The character the input uses to denote the decimal/fractional portion of the number. Defaults to period `.`",
+      type: "string",
+      options: [
```
Comment (vellames) [inline components/formatting/actions/format-number/format-number.ts:23]: Could you please move the `options` of this file to a separated `options` file?

File: components/formatting/actions/split-text/split-text.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      description:
+        "Character or word separator to split the text on. This can also be a string representing a [Regular Expression](https://www.w3schools.com/js/js_regexp.asp)",
+      type: "string",
+    },
+    segmentIndex: {
+      label: "Segment Index",
+      type: "integer",
+      description:
+        "Segment of text to return after splitting. Choose one of the options, or use a custom positive (*nth*-match) or negative (*nth-to-last* match) integer.",
+      options: [
```
Comment (vellames) [inline components/formatting/actions/split-text/split-text.ts:30]: Could you please move these `options` to an options file?

File: components/formatting/actions/transform-case/transform-case.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    input: {
+      label: "Input",
+      description: "The text you would like to transform",
+      type: "string",
+    },
+    operation: {
+      label: "Operation",
+      description: "The case operation",
+      type: "string",
+      options: CASE_OPTIONS.map(({
```
Comment (vellames) [inline components/formatting/actions/transform-case/transform-case.ts:20]: Could you please move these `options` to an options file?

**Review by vellames — CHANGES_REQUESTED** [review]

**Review by vellames — CHANGES_REQUESTED** [review]
Lint problems :/
Please check pipeline

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[Fomatting_5172.pdf](https://github.com/PipedreamHQ/pipedream/files/10707592/Fomatting_5172.pdf)

**Review by vellames — APPROVED** [review]

**Pull request merged**